File: backend/src/schemas/me.py
from pydantic import BaseModel, Field


class UpdateProfileRequest(BaseModel):
    display_name: str | None = Field(default=None, max_length=128)
    age: int | None = Field(default=None, ge=0, le=150)
    color: str | None = Field(default=None, max_length=16)
    bio: str | None = Field(default=None, max_length=2000)
    favo: str | None = Field(default=None, max_length=2000)


class ChangePasswordRequest(BaseModel):
    current_password: str
    new_password: str = Field(min_length=4, max_length=128)

File: frontend/src/api/me.ts
import { apiFetch } from "./client";
import type { UserResponse } from "./auth";

export interface UpdateProfileRequest {
  display_name?: string | null;
  age?: number | null;
  color?: string | null;
  bio?: string | null;
  favo?: string | null;
}

export function updateProfile(payload: UpdateProfileRequest): Promise<UserResponse> {
  return apiFetch<UserResponse>("/api/me", { method: "PATCH", body: payload });
}

export function changePassword(currentPassword: string, newPassword: string): Promise<void> {
  return apiFetch<void>("/api/me/password", {
    method: "POST",
    body: { current_password: currentPassword, new_password: newPassword },
  });
}

export async function uploadAvatar(file: File): Promise<UserResponse> {
  const form = new FormData();
  form.append("file", file);
  const res = await fetch("/api/me/avatar", {
    method: "POST",
    credentials: "include",
    body: form,
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(text || `Upload failed (${res.status})`);
  }
  return (await res.json()) as UserResponse;
}
